Modules/_io/fileio.c

```diff
@@ -707,41 +712,48 @@ static PyObject *
 _io_FileIO_readall_impl(fileio *self)
 /*[clinic end generated code: output=faa0292b213b4022 input=dbdc137f55602834]*/
 {
-    struct _Py_stat_struct status;
     Py_off_t pos, end;
     PyObject *result;
     Py_ssize_t bytes_read = 0;
     Py_ssize_t n;
     size_t bufsize;
-    int fstat_result;
 
-    if (self->fd < 0)
+    if (self->fd < 0) {
         return err_closed();
+    }
 
-    Py_BEGIN_ALLOW_THREADS
-    _Py_BEGIN_SUPPRESS_IPH
+    end = self->size_estimated;
+    if (end <= 0) {
+        /* Use a default size and resize as needed. */
+        bufsize = SMALLCHUNK;
+    }
+    else {
+        /* This is probably a real file, so we try to allocate a
+        buffer one byte larger than the rest of the file.  If the
+        calculation is right then we should get EOF without having
+        to enlarge the buffer. */
+        bufsize = (size_t)(end) + 1;
+
+        /* While a lot of code does open().read() to get the whole contents
+        of a file it is possible a caller seeks/reads a ways into the file
+        then calls readall() to get the rest, which would result in allocating
+        more than required. Guard against that for larger files where we expect
+        the I/O time to dominate anyways while keeping small files fast. */
+        if (bufsize > 65536) {
```

Comment:
I don't like having a hardcoded constant like that. I'd suggest making a named macro.